<a href="https://colab.research.google.com/github/jacobpad/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Sprint_1_Module_1_Assignment_Review.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-01-21 16:09:03--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-01-21 16:09:04 (132 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
# Imports
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import seaborn as sns

In [0]:
df = pd.read_csv('house-votes-84.data', 
                 header=None,
                 names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime',
                          'duty-free', 'south-africa'],
                 na_values="?")
print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
df = df.replace({'y':1, 'n':0})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [0]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

In [0]:
267+168

435

In [0]:
# One way to replace question marks with NaNs
# df = df.replace({"?": np.NaN})

In [0]:
dem = df[df['party'] == "democrat"]

In [0]:
rep = df[df['party'] == "republican"]

In [0]:
# 88.8% of democrats voted "YES" on the budget bill
dem['budget'].mean()

0.8884615384615384

# One Sample T-test (Democrats on the budget bill)

Null Hypothesis: 

In one sample t-tests we get to pick whatever value we want, for our null hypothesis. We will be comparing our sample mean to this value, and asking - Are they the same.

$H_0 == .5$

I want to ask the question: Is the proportion of democrats who support the budget bill, statistically different than .5 - Meaning, the democrats are split in their support for the budget bill (50% yes, 50% no). 

Are the democrats split in their support for this bill, or is their support level something *other than* 50%. 

Alternative Hypothesis: (opposite of the null hypothesis)

$H_a \neq .5$

Confidencd Level: 95% Confidence Level

In [0]:
# ttest_ind (two sample t-tests)
# ttest_1samp (one sample t-tests)

### Equation for the t statistic in a 1-sample t-test
![t stat equation](https://benmis.files.wordpress.com/2013/10/t_test_formula_one_way.gif)

In [0]:
dem['budget'].value_counts(dropna=False)

1.0    231
0.0     29
NaN      7
Name: budget, dtype: int64

In [0]:
numerator =  (dem['budget'].mean() - .5) 
print("Diference betweeen the null hypothesis value and the sample mean:", numerator)

Diference betweeen the null hypothesis value and the sample mean: 0.3884615384615384


In [0]:
s = dem['budget'].std()
sqrt_n = np.sqrt(260)

denomenator = s / sqrt_n

standard_error = denomenator

In [0]:
t = numerator / standard_error

t

19.859406568628863

## Standard Error changes with sample size

In [0]:
# ten coinflips
coinflips = np.random.binomial(n=1, p=.5, size=10)

coinflips.mean()

0.4

In [0]:
# one hundred coinflips
coinflips = np.random.binomial(n=1, p=.5, size=100)

coinflips.mean()

0.5002381

## The Standard Error Quantifies the randomness of an estimate

Largely driven by our sample size. 

Larger Sample size ==> Less random estimates (closer to population mean)

Smaller Sample size ==> More random estimates (can vary more from the population mean)

In [0]:
stats.ttest_1samp(dem['budget'], .5, nan_policy='omit')

Ttest_1sampResult(statistic=19.859406568628835, pvalue=5.75931504660857e-54)

P-value threshold: (1 - confidence_level) == .05

a p-value that is less than .05 is statistically significant

Statistical Significance means: Our sample mean is different enough from the null hypothesis value, that it is unlikely to have happened due to random chance (unluckiness)

What is the probability here that this is due to random chance? - p-value

P-value < .05 => Reject the null hypothesis

## Conclusion

Based on a p-value of 0 and a t-statistic of 19.85, I reject the null hypothesis that democrat support for the budget bill is 50%, and suggest the alternative, that democrat support is not 50%. (democrats are not split on this issue)

# Two-sample T-test

I don't get to choose the null hypothesis value here.

I let the mean of the second sample serve as the value that I'm comparing the first sample's mean to. Comparing two sample means directly.

$ H_0: \text{dem_mean} == \text{rep_mean}$ 

$ H_a: \text{dem_mean} \neq \text{rep_mean}$ 

Confidence_Level = 95%



In [0]:
stats.ttest_ind(dem['budget'], rep['budget'], nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

Based on a p-value of 2.07x10^-77 and a test statistic of 23.2, I **reject the null hypothesis** that democrat support for the budget bill is equal to republican support for the budget bill, and I suggest the alternative that the support for this bill between the two parties is different. 

In [0]:
dem['budget'].mean()

0.8884615384615384

In [0]:
rep['budget'].mean()

0.13414634146341464